src/main/java/org/openqa/selenium/remote/NewSessionPayload.java

```diff
@@ -0,0 +1,580 @@
+            .addAll(getAppiumCapabilities(IOSMobileCapabilityType.class))
+            .addAll(getAppiumCapabilities(YouiEngineCapabilityType.class)).build();
+
+    private final Set<CapabilitiesFilter> adapters;
+    private final Set<CapabilityTransform> transforms;
+    private final boolean forceMobileJSONWP;
+
+    private static final Dialect DEFAULT_DIALECT = Dialect.OSS;
+    private final static Predicate<String> ACCEPTED_W3C_PATTERNS = Stream.of(
+            "^[\\w-]+:.*$",
+            "^acceptInsecureCerts$",
+            "^browserName$",
+            "^browserVersion$",
+            "^platformName$",
+            "^pageLoadStrategy$",
+            "^proxy$",
+            "^setWindowRect$",
+            "^timeouts$",
+            "^unhandledPromptBehavior$")
+            .map(Pattern::compile)
+            .map(Pattern::asPredicate)
+            .reduce(identity -> false, Predicate::or);
+
+    private final Json json = new Json();
+    private final FileBackedOutputStream backingStore;
+    private final ImmutableSet<Dialect> dialects;
+
+    private static List<String> getAppiumCapabilities(Class<?> capabilityList) {
+        return Arrays.stream(capabilityList.getDeclaredFields()).map(field -> {
+                    field.setAccessible(true);
+            try {
+                return field.get(capabilityList).toString();
+            } catch (IllegalAccessException e) {
+                throw new IllegalArgumentException(e);
+            }
+        }).filter(s -> !FORCE_MJSONWP.equals(s)).collect(toList());
+    }
+
+    public static NewSessionPayload create(Capabilities caps) throws IOException {
+        // We need to convert the capabilities into a new session payload. At this point we're dealing
+        // with references, so I'm Just Sure This Will Be Fine.
+        boolean forceMobileJSONWP = ofNullable(caps.getCapability(FORCE_MJSONWP))
+                .map(o -> {
+                    if (Boolean.class.isAssignableFrom(o.getClass())) {
+                        return Boolean.class.cast(o);
+                    }
+                    return false;
+                }).orElse(false);
+        HashMap<String, ?> capabilityMap = new HashMap<>(caps.asMap());
+        capabilityMap.remove(FORCE_MJSONWP);
+        return create(ImmutableMap.of("desiredCapabilities", capabilityMap), forceMobileJSONWP);
+    }
+
+    public static NewSessionPayload create(Map<String, ?> source, boolean forceMobileJSONWP) throws IOException {
+        Objects.requireNonNull(source, "Payload must be set");
+
+        String json = new Json().toJson(source);
+        return new NewSessionPayload(new StringReader(json), forceMobileJSONWP);
+    }
+
+    public static NewSessionPayload create(Reader source, boolean forceMobileJSONWP) throws IOException {
+        return new NewSessionPayload(source, forceMobileJSONWP);
+    }
+
+    private NewSessionPayload(Reader source, boolean forceMobileJSONWP) throws IOException {
```

Comment:
This copies a lot of stuff from the original selenium code. Is it somehow possible to avoid such duplication?

Reply:
for example using reflection